src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs

```diff
@@ -1,38 +1,212 @@
+			}
+
+			// When having multiple level closure, we have to evaluate each closure independently
+			if (evaluatedExpression is ConstantExpression constantExpression)
+			{
+				evaluatedExpression = VisitConstant(constantExpression);
+			}
+
+			// Variables in expressions are never a constant, they are encapsulated as fields of a compiler generated class
+			if (expression.NodeType != ExpressionType.Constant &&
+			    evaluatedExpression is ConstantExpression variableConstant &&
+			    !_queryVariables.ContainsKey(variableConstant) &&
+			    IsVariable(expression, out var path, out var closureContext))
+			{
+				_queryVariables.Add(variableConstant, new QueryVariable(path, closureContext));
+			}
+
+			return evaluatedExpression;
+		}
+
 		protected override Expression VisitConstant(ConstantExpression expression)
 		{
 			if (expression.Value is Expression value)
 			{
-				return EvaluateIndependentSubtrees(value);
+				return EvaluateIndependentSubtrees(value, _evaluatableExpressionFilter, _queryVariables);
 			}
 
 			return base.VisitConstant(expression);
 		}
 
-		public static Expression EvaluateIndependentSubtrees(Expression expression)
+		/// <summary>
+		/// Evaluates an evaluatable <see cref="Expression"/> subtree, i.e. an independent expression tree that is compilable and executable
+		/// without any data being passed in. The result of the evaluation is returned as a <see cref="ConstantExpression"/>; if the subtree
+		/// is already a <see cref="ConstantExpression"/>, no evaluation is performed.
+		/// </summary>
+		/// <param name="subtree">The subtree to be evaluated.</param>
+		/// <returns>A <see cref="ConstantExpression"/> holding the result of the evaluation.</returns>
+		private Expression EvaluateSubtree(Expression subtree)
+		{
+			if (subtree.NodeType == ExpressionType.Constant)
+			{
+				var constantExpression = (ConstantExpression) subtree;
+				var valueAsIQueryable = constantExpression.Value as IQueryable;
+				if (valueAsIQueryable != null && valueAsIQueryable.Expression != constantExpression)
+					return valueAsIQueryable.Expression;
+
+				return constantExpression;
+			}
+			else
+			{
+				Expression<Func<object>> lambdaWithoutParameters = Expression.Lambda<Func<object>>(Expression.Convert(subtree, typeof(object)));
+				var compiledLambda = lambdaWithoutParameters.Compile();
+
+				object value = compiledLambda();
+				return Expression.Constant(value, subtree.Type);
+			}
+		}
+
+		private bool IsVariable(Expression expression, out string path, out object closureContext)
+		{
+			Expression childExpression;
+			string currentPath;
+			switch (expression)
+			{
+				case MemberExpression memberExpression:
+					childExpression = memberExpression.Expression;
+					currentPath = memberExpression.Member.Name;
+					break;
+				case ConstantExpression constantExpression:
+					path = null;
+					if (constantExpression.Type.Attributes.HasFlag(TypeAttributes.NestedPrivate) &&
```

Comment:
Used the same condition as Microsoft uses in [their ORM](https://github.com/dotnet/efcore/blob/e400969114d7a671b46bd9ba7b903bd8e7d63329/src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs#L604) in order to detect the compiler generated class containing the variables.